include/experimental/__p2642_bits/layout_padded.hpp

```diff
@@ -0,0 +1,808 @@
+  static constexpr size_t padding_value = PaddingValue;
+
+  using index_type = typename extents_type::index_type;
+  using size_type = typename extents_type::size_type;
+  using rank_type = typename extents_type::rank_type;
+  using layout_type = layout_left_padded<padding_value>;
+
+#ifndef MDSPAN_INTERNAL_TEST
+private:
+#endif // MDSPAN_INTERNAL_TEST
+
+  static constexpr rank_type __padded_stride_idx = detail::__layout_padded_constants<layout_type, extents_type>::__padded_stride_idx;
+  static constexpr rank_type __extent_to_pad_idx = detail::__layout_padded_constants<layout_type, extents_type>::__extent_to_pad_idx;
+
+  static_assert((padding_value != 0)
+                || (extents_type::static_extent(__extent_to_pad_idx) == 0)
+                || (extents_type::static_extent(__extent_to_pad_idx) == dynamic_extent),
+                "out of bounds access for rank 0");
+
+  static constexpr size_t __actual_padding_value = detail::__get_actual_static_padding_value<extents_type, padding_value, __extent_to_pad_idx>();
+
+  using __padded_stride_type = detail::__padded_extent< padding_value, extents_type, __extent_to_pad_idx >;
+
+  typename __padded_stride_type::__static_array_type __padded_stride = {};
+  extents_type __extents = {};
+
+  constexpr index_type __compute_offset(std::index_sequence<>) const {
+    return 0;
+  }
+
+  template <size_t Rank, class IndexOffset>
+  constexpr index_type __compute_offset(std::index_sequence<Rank>,
+                                        IndexOffset __index_offset) const {
+    return __index_offset;
+  }
+
+  template <size_t... Ranks, class... IndexOffsets>
+  constexpr index_type __compute_offset(std::index_sequence<Ranks...>,
+                                        IndexOffsets... __index_offsets) const {
+    index_type __indices[] = {static_cast<index_type>(__index_offsets)...};
+    // self-recursive fold trick from
+    // https://github.com/llvm/llvm-project/blob/96e1914aa2e6d8966acbfbe2f4d184201f1aa318/libcxx/include/__mdspan/layout_left.h#L144
+    index_type res = 0;
+    ((res = __indices[extents_type::rank() - 1 - Ranks] +
+            ((extents_type::rank() - 1 - Ranks) == __extent_to_pad_idx
+                 ? __padded_stride.value(0)
+                 : __extents.extent(extents_type::rank() - 1 - Ranks)) *
+                res),
+     ...);
+    return res;
+  }
+
+public:
+#if !MDSPAN_HAS_CXX_20
+  MDSPAN_INLINE_FUNCTION_DEFAULTED
+  constexpr mapping()
+      : mapping(extents_type{})
+  {}
+#else
+  MDSPAN_INLINE_FUNCTION_DEFAULTED
+  constexpr mapping()
+    requires(__actual_padding_value != dynamic_extent) = default;
+
+  MDSPAN_INLINE_FUNCTION
+  constexpr mapping()
+    requires(__actual_padding_value == dynamic_extent)
+      : mapping(extents_type{})
+  {}
+#endif
+
+  MDSPAN_INLINE_FUNCTION_DEFAULTED constexpr mapping(const mapping&) noexcept = default;
+  MDSPAN_INLINE_FUNCTION_DEFAULTED mapping& operator=(const mapping&) noexcept = default;
```

Comment:
I see them in the wording but I can't help wondering why do we bother?

Reply:
Declaring the copy constructor and copy assignment operator explicitly, but not their corresponding move special members, makes the class non-moveable but copyable.  LWG really seemed to want that.  I think the point is to make clear that there is no moved-from state to worry about.  (On the other hand, that should be obvious from the mapping only containing `extents` and things like it.)

Reply:
https://godbolt.org/z/KG487sh81

Reply:
To elaborate:  Because move is an optimization of copy, types that are copyable are always moveable as they fall back on calling the copy operation.

For instance, it is why declaring member variables as `const` or using references can be a pessimization, because it deletes the move assignment operator (even if defaulted), and the attempt to move the object just copies it.

(FYI: this is subtle and surprising.)

Unrelated question:  why isn't the assignment operator `constexpr` (it isn't in the paper either)?

Reply:
My point was why do we explicitly default them instead of not writing anything and letting the compiler do its job.
```suggestion
```

Reply:
because the standard puts them explicitly in too for some reason?